summingbird-core/src/main/scala/com/twitter/summingbird/memory/Memory.scala

```diff
@@ -39,14 +41,14 @@ class Memory(implicit jobID: JobId = JobId("default.memory.jobId")) extends Plat
   type Plan[T] = Stream[T]
 
   private type Prod[T] = Producer[Memory, T]
-  private type JamfMap = HMap[Prod, Stream]
+  private type JamfMap = util.IdentityHashMap[Prod[_], Stream[_]]
```

Comment:
this is mutable now, and before it was immutable. Does this not cause any problems? Can you add some comments explaining how it is safe? Also, why thread it through if you are mutating it? There is no need to return `JamfMap` if you are just returning the input.

I would prefer not introduce a mutable map, and instead add:
```scala
class Identity[T](val unwrap: T) {
  override def equals(that: Any) = that match {
    case i: Identity[_] => unwrap eq i.unwrap
    case _ => false
  }
  override def hashCode = System.identityHashCode(unwrap)
}
private type Prod[T] = Identity[Producer[Memory, T]]
```
then wrap keys with `new Identity(key)`.

I really hate to give up the reasoning we get with immutable types, and we should only do so for big performance wins, but we don't care about performance here (planning is fast, and only happens at submit).